whois-nrtm4-client/src/main/java/net/ripe/db/nrtm4/client/importer/SnapshotImporter.java

```diff
@@ -0,0 +1,170 @@
+            //  call initialize X number of times and return error to avoid this situation?
+            LOGGER.error("The session is not the same in the UNF and snapshot");
+            //initializeNRTMClientForSource(source, updateNotificationFile);
+            return;
+        }
+
+        printProgress(new Timer());
+        Arrays.stream(snapshotRecords).skip(1)
+                .parallel()
+                .forEach(record -> {
+                    try {
+                        processObject(record);
+                        processedCount.incrementAndGet();
+                    } catch (JsonProcessingException e) {
+                        LOGGER.error("Unable to process record", e);
+                    }
+                });
+
+        nrtm4ClientMirrorDao.saveSnapshotFileVersion(source, snapshotVersion, snapshotSessionId);
+        LOGGER.info("Loading snapshot file took {} for source {}", stopwatch.elapsed().toMillis(), source);
+    }
+
+    private void processObject(final String record) throws JsonProcessingException {
+        final MirrorRpslObject mirrorRpslObject = new ObjectMapper().readValue(record, MirrorRpslObject.class);
+        nrtm4ClientMirrorDao.persistRpslObject(mirrorRpslObject.getObject());
+    }
+
+    private void printProgress(final Timer timer) {
+        timer.schedule(new TimerTask() {
+            @Override
+            public void run() {
+                LOGGER.info("Records processed so far: {}", processedCount.get());
+            }
+        }, 0, 10000);
+    }
+
+    private static String[] getSnapshotRecords(byte[] compressed) throws IOException {
+        return StringUtils.split(decompress(compressed), RECORD_SEPARATOR);
+    }
+
+    private static String decompress(final byte[] compressed) throws IOException {
+        final int BUFFER_SIZE = 4096;
+        try (ByteArrayInputStream is = new ByteArrayInputStream(compressed);
+             GZIPInputStream gis = new GZIPInputStream(is, BUFFER_SIZE);
+             ByteArrayOutputStream output = new ByteArrayOutputStream()) {
+            byte[] buffer = new byte[BUFFER_SIZE];
+            int bytesRead;
+            while ((bytesRead = gis.read(buffer)) != -1) {
+                output.write(buffer, 0, bytesRead);
+            }
+            return output.toString(StandardCharsets.UTF_8);
+        }
+    }
+
+    private static String calculateSha256(final byte[] bytes) {
+        try {
+            final MessageDigest digest = MessageDigest.getInstance("SHA-256");
+            final byte[] encodedSha256hex = digest.digest(bytes);
+            return byteArrayToHexString(encodedSha256hex);
+        } catch (final NoSuchAlgorithmException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static String byteArrayToHexString(final byte[] bytes) {
+        final StringBuilder hexString = new StringBuilder(2 * bytes.length);
```

Comment:
Check the consequences of allocating a new `StringBuilder` that's double the size of the snapshot (which is 100's of megabytes). Maybe there is a more efficient way of calculating the SHA hash ?

Reply:
I did refactor this. I found a better solution related to performance.
I pushed the change

Reply:
I will try with the library org.apache.commons.codec.binary.Hex.encodeHexString(byte[])
That you pointed out. I am doing on this way because we are doing the same in IT for Nrtm v4